lib/Onyx.d.ts

```diff
@@ -0,0 +1,315 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {
+    CollectionKey,
+    CollectionKeyBase,
+    DeepRecord,
+    KeyValueMapping,
+    OnyxCollection,
+    OnyxEntry,
+    OnyxKey,
+} from './types';
+
+/**
+ * Represents a mapping object where each `OnyxKey` maps to either a value of its corresponding type in `KeyValueMapping` or `null`.
+ *
+ * It's very similar to `KeyValueMapping` but this type accepts using `null` as well.
+ */
+type NullableKeyValueMapping = {
+    [TKey in OnyxKey]: OnyxEntry<KeyValueMapping[TKey]>;
+};
+
+/**
+ * Represents the base options used in `Onyx.connect()` method.
+ */
+type BaseConnectOptions = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+};
+
+/**
+ * Represents the options used in `Onyx.connect()` method.
+ * The type is built from `BaseConnectOptions` and extended to handle key/callback related options.
+ * It includes two different forms, depending on whether we are waiting for a collection callback or not.
+ *
+ * If `waitForCollectionCallback` is `true`, it expects `key` to be a Onyx collection key and `callback` will pass `value` as an `OnyxCollection`.
+ *
+ * If `waitForCollectionCallback` is `false` or not specified, the `key` can be any Onyx key and `callback` will pass `value` as an `OnyxEntry`.
```

Comment:
@fabioh8010 can you explain that if the Onyx key is a collection key and `waitForCollectionCallback` is either not specified or `false`, the callback will be triggered for updates of each collection item?

I know you didn't design the API, but this was really confusing to me so a comment would be helpful 🙇

Reply:
@hayata-suenaga Adjusted!